File: src/models/gdcn_esmm.rs
//! GDCN + ESMM: gated cross network shared representation with ESMM task towers.
use super::output_contract::OutputContract;
use super::output_head::OutputHead;
use super::{Model, ModelExecution, ModelOutput};
use crate::layers::embedding::{FeatureEmbeddings, FeatureSpec};
use crate::layers::gdcn::GatedCrossNetwork;
use crate::layers::mlp::Mlp;
use crate::layers::towers::{apply_relation, Activation, MultiTaskConfig, TaskRelation, TaskTower};
use crate::models::esmm::default_task_config;
use candle_core::{Result, Tensor};
use candle_nn::{Module, VarBuilder};
use std::collections::HashMap;

/// ESMM variant using a gated cross network plus optional deep branch.
pub struct GDCNESMM {
    embeddings: FeatureEmbeddings,
    cross: GatedCrossNetwork,
    deep: Option<Mlp>,
    shared_bottom: Option<Mlp>,
    towers: Vec<(String, TaskTower)>,
    relations: Vec<TaskRelation>,
    output_head: Option<OutputHead>,
}

impl GDCNESMM {
    #[allow(clippy::too_many_arguments)]
    /// 使用默认任务配置构造 GDCNESMM。
    pub fn new(
        vb: VarBuilder,
        features: &[FeatureSpec],
        cross_layers: usize,
        deep_hidden_dims: &[usize],
        shared_bottom_dims: &[usize],
        click_hidden_dims: &[usize],
        cvr_hidden_dims: &[usize],
        detail_hidden_dims: &[usize],
        stock_hidden_dims: &[usize],
        stay_hidden_dims: &[usize],
    ) -> Result<Self> {
        let task_config = default_task_config(
            click_hidden_dims,
            cvr_hidden_dims,
            detail_hidden_dims,
            stock_hidden_dims,
            stay_hidden_dims,
        );
        Self::with_task_config(
            vb,
            features,
            cross_layers,
            deep_hidden_dims,
            shared_bottom_dims,
            &task_config,
        )
    }

    /// 使用自定义 MultiTaskConfig 构造 GDCNESMM。
    pub fn with_task_config(
        vb: VarBuilder,
        features: &[FeatureSpec],
        cross_layers: usize,
        deep_hidden_dims: &[usize],
        shared_bottom_dims: &[usize],
        task_config: &MultiTaskConfig,
    ) -> Result<Self> {
        let embeddings = FeatureEmbeddings::new(vb.pp("embeddings"), features)?;
        let input_dim = embeddings.total_dim;
        let cross = GatedCrossNetwork::new(vb.pp("cross"), input_dim, cross_layers)?;

        let (deep, fusion_dim) = if deep_hidden_dims.is_empty() {
            (None, input_dim)
        } else {
            let output_dim = *deep_hidden_dims.last().unwrap();
            let mlp = Mlp::new(
                vb.pp("deep"),
                input_dim,
                &deep_hidden_dims[..deep_hidden_dims.len() - 1],
                output_dim,
                Activation::Relu,
            )?;
            (Some(mlp), input_dim + output_dim)
        };

        let (shared_bottom, tower_input_dim) = if shared_bottom_dims.is_empty() {
            (None, fusion_dim)
        } else {
            let output_dim = *shared_bottom_dims.last().unwrap();
            let mlp = Mlp::new(
                vb.pp("shared_bottom"),
                fusion_dim,
                &shared_bottom_dims[..shared_bottom_dims.len() - 1],
                output_dim,
                Activation::Relu,
            )?;
            (Some(mlp), output_dim)
        };

        let mut towers = Vec::with_capacity(task_config.towers.len());
        for tower_config in &task_config.towers {
            towers.push((
                tower_config.name.clone(),
                TaskTower::new(
                    tower_config,
                    tower_input_dim,
                    vb.pp(format!("{}_tower", tower_config.name)),
                )?,
            ));
        }

        Ok(Self {
            embeddings,
            cross,
            deep,
            shared_bottom,
            towers,
            relations: task_config.relations.clone(),
            output_head: None,
        })
    }

    /// 使用原生输出契约构造 GDCNESMM。
    pub fn with_output_contract(
        vb: VarBuilder,
        features: &[FeatureSpec],
        cross_layers: usize,
        deep_hidden_dims: &[usize],
        shared_bottom_dims: &[usize],
        contract: &OutputContract,
    ) -> Result<Self> {
        let embeddings = FeatureEmbeddings::new(vb.pp("embeddings"), features)?;
        let input_dim = embeddings.total_dim;
        let cross = GatedCrossNetwork::new(vb.pp("cross"), input_dim, cross_layers)?;
        let (deep, fusion_dim) = if deep_hidden_dims.is_empty() {
            (None, input_dim)
        } else {
            let output_dim = *deep_hidden_dims.last().unwrap();
            (
                Some(Mlp::new(
                    vb.pp("deep"),
                    input_dim,
                    &deep_hidden_dims[..deep_hidden_dims.len() - 1],
                    output_dim,
                    Activation::Relu,
                )?),
                input_dim + output_dim,
            )
        };
        let (shared_bottom, shared_dim) = if shared_bottom_dims.is_empty() {
            (None, fusion_dim)
        } else {
            let output_dim = *shared_bottom_dims.last().unwrap();
            (
                Some(Mlp::new(
                    vb.pp("shared_bottom"),
                    fusion_dim,
                    &shared_bottom_dims[..shared_bottom_dims.len() - 1],
                    output_dim,
                    Activation::Relu,
                )?),
                output_dim,
            )
        };
        let output_head = OutputHead::new(
            contract,
            &HashMap::from([("shared".to_string(), shared_dim)]),
            vb.pp("output_head"),
        )?;
        Ok(Self {
            embeddings,
            cross,
            deep,
            shared_bottom,
            towers: vec![],
            relations: vec![],
            output_head: Some(output_head),
        })
    }

    fn shared(&self, x_inputs: &HashMap<String, Tensor>) -> Result<Tensor> {
        let dense = self.embeddings.forward(x_inputs)?;
        let cross_out = self.cross.forward(&dense)?;
        let mut shared = match &self.deep {
            Some(deep) => Tensor::cat(&[cross_out, deep.forward(&dense)?], 1)?,
            None => cross_out,
        };
        if let Some(shared_bottom) = &self.shared_bottom {
            shared = shared_bottom.forward(&shared)?;
        }
        Ok(shared)
    }
}

impl Model for GDCNESMM {
    fn forward(&self, x_inputs: &HashMap<String, Tensor>) -> Result<ModelOutput> {
        if self.output_head.is_some() {
            return Ok(self.forward_execution(x_inputs)?.outputs);
        }
        let shared = self.shared(x_inputs)?;
        let mut outputs = ModelOutput::new();
        for (name, tower) in &self.towers {
            outputs.insert_binary_logit(name.clone(), tower.forward(&shared)?);
        }
        for relation in &self.relations {
            outputs
                .insert_probability(relation.target.clone(), apply_relation(relation, &outputs)?);
        }
        Ok(outputs)
    }

    fn forward_execution(&self, x_inputs: &HashMap<String, Tensor>) -> Result<ModelExecution> {
        if let Some(head) = &self.output_head {
            return head.forward(&HashMap::from([(
                "shared".to_string(),
                self.shared(x_inputs)?,
            )]));
        }
        let outputs = self.forward(x_inputs)?;
        Ok(ModelExecution::new(outputs.clone(), outputs))
    }
}

File: python/src/train/models/gdcn_esmm.py
from __future__ import annotations

"""GDCN + ESMM: gated cross network shared representation with ESMM task towers."""

from typing import TYPE_CHECKING

import torch
import torch.nn as nn

if TYPE_CHECKING:
    from ..core.config import PoolingMode

from ..core.model_output import ModelExecution, ModelOutput
from ..core.output_contract import NormalizedOutputContract
from ..layers.embedding import FeatureEmbeddings, FeatureTensorMap, FeatureTuple
from ..layers.gdcn import GatedCrossNetwork
from ..layers.mlp import Mlp
from ..layers.towers import Activation, MultiTaskConfig, TaskRelation, TaskTower
from .esmm import _probability_for_relation, default_task_config
from .output_head import OutputHead


class GDCNESMM(nn.Module):
    """ESMM variant using a gated cross network plus optional deep branch."""

    def __init__(
        self,
        features: list[FeatureTuple],
        cross_layers: int = 3,
        deep_hidden_dims: list[int] | None = None,
        shared_bottom_dims: list[int] | None = None,
        click_hidden_dims: list[int] | None = None,
        cvr_hidden_dims: list[int] | None = None,
        detail_hidden_dims: list[int] | None = None,
        stock_hidden_dims: list[int] | None = None,
        stay_hidden_dims: list[int] | None = None,
        task_config: MultiTaskConfig | None = None,
        pooling_map: dict[str, PoolingMode] | None = None,
        total_dim: int | None = None,
        output_contract: NormalizedOutputContract | None = None,
    ) -> None:
        super().__init__()
        deep_hidden_dims = deep_hidden_dims or []
        shared_bottom_dims = shared_bottom_dims or []
        click_hidden_dims = click_hidden_dims or [8]
        cvr_hidden_dims = cvr_hidden_dims or [8]
        detail_hidden_dims = detail_hidden_dims or [8]
        stock_hidden_dims = stock_hidden_dims or [8]
        stay_hidden_dims = stay_hidden_dims or [8]

        self.embeddings = FeatureEmbeddings(features, pooling_map, total_dim=total_dim)
        input_dim = self.embeddings.total_dim
        self.cross = GatedCrossNetwork(input_dim, cross_layers)
        self.has_deep = bool(deep_hidden_dims)
        if self.has_deep:
            self.deep = Mlp(input_dim, deep_hidden_dims[:-1], deep_hidden_dims[-1], Activation.RELU)
            fusion_dim = input_dim + deep_hidden_dims[-1]
        else:
            fusion_dim = input_dim

        if shared_bottom_dims:
            self.shared_bottom = Mlp(
                fusion_dim,
                shared_bottom_dims[:-1],
                shared_bottom_dims[-1],
                Activation.RELU,
            )
            tower_input_dim = shared_bottom_dims[-1]
        else:
            tower_input_dim = fusion_dim

        if output_contract is not None:
            self.output_head = OutputHead(output_contract, {"shared": tower_input_dim})
            self.task_names = [tower.name for tower in output_contract.towers]
            self.relation_names = list(output_contract.relation_order)
            return

        self.task_config = task_config or default_task_config(
            click_hidden_dims,
            cvr_hidden_dims,
            detail_hidden_dims,
            stock_hidden_dims,
            stay_hidden_dims,
        )
        self.task_names = [tower.name for tower in self.task_config.towers]
        self.relation_names = [relation.target for relation in self.task_config.relations]
        for tower in self.task_config.towers:
            setattr(self, f"{tower.name}_tower", TaskTower(tower, tower_input_dim))

    def forward(self, x_inputs: FeatureTensorMap) -> ModelOutput:
        if hasattr(self, "output_head"):
            return self.forward_execution(x_inputs).outputs
        return self._forward_legacy(self._shared(x_inputs))

    def forward_execution(self, x_inputs: FeatureTensorMap) -> ModelExecution:
        shared = self._shared(x_inputs)
        if hasattr(self, "output_head"):
            return self.output_head({"shared": shared})
        outputs = self._forward_legacy(shared)
        return ModelExecution(nodes=outputs, outputs=outputs)

    def _shared(self, x_inputs: FeatureTensorMap) -> torch.Tensor:
        dense = self.embeddings(x_inputs)
        cross_out = self.cross(dense)
        shared = torch.cat([cross_out, self.deep(dense)], dim=1) if self.has_deep else cross_out
        if hasattr(self, "shared_bottom"):
            shared = self.shared_bottom(shared)
        return shared

    def _forward_legacy(self, shared: torch.Tensor) -> ModelOutput:
        outputs = ModelOutput()
        for name in self.task_names:
            tower = getattr(self, f"{name}_tower")
            outputs.insert(name, tower(shared), tower.output_kind)
        for relation in self.task_config.relations:
            outputs.insert_probability(relation.target, self._apply_relation(relation, outputs))
        return outputs

    @staticmethod
    def _apply_relation(relation: TaskRelation, outputs: ModelOutput) -> torch.Tensor:
        if not relation.sources:
            raise ValueError(f"Relation '{relation.target}' has no sources")
        probs = [
            _probability_for_relation(relation, outputs, source) for source in relation.sources
        ]
        if relation.op == "multiply":
            result = probs[0]
            for value in probs[1:]:
                result = result * value
            return result
        if relation.op == "add":
            result = probs[0]
            for value in probs[1:]:
                result = result + value
            return result
        if relation.op == "subtract":
            if len(probs) != 2:
                raise ValueError(f"Relation '{relation.target}' subtract requires 2 sources")
            return probs[0] - probs[1]
        if relation.op == "divide":
            if len(probs) != 2:
                raise ValueError(f"Relation '{relation.target}' divide requires 2 sources")
            return probs[0] / (probs[1] + 1e-8)
        raise ValueError(f"Unknown relation op: {relation.op}")
